File: org.summer.view.widget/src/org/summer/view/widget/collection/ReadOnlyObservableCollection.java
package org.summer.view.widget.collection;

import org.summer.view.widget.INotifyCollectionChanged;
import org.summer.view.widget.INotifyPropertyChanged;
import org.summer.view.widget.model.PropertyChangedEventArgs;

/// <summary> 
/// Read-only wrapper around an ObservableCollection.
/// </summary> 
//[Serializable()] 
//[TypeForwardedFrom("WindowsBase, Version=3.0.0.0, Culture=Neutral, PublicKeyToken=31bf3856ad364e35")]
public class ReadOnlyObservableCollection<T> extends ReadOnlyCollection<T> implements INotifyCollectionChanged, INotifyPropertyChanged 
{
//    #region Constructors

    //----------------------------------------------------- 
    //
    //  Constructors 
    // 
    //-----------------------------------------------------

    /// <summary>
    /// Initializes a new instance of ReadOnlyObservableCollection that
    /// wraps the given ObservableCollection.
    /// </summary> 
    public ReadOnlyObservableCollection(ObservableCollection<T> list)
    { 
    	 super(list);
        ((INotifyCollectionChanged)Items).CollectionChanged += new NotifyCollectionChangedEventHandler(HandleCollectionChanged); 
        ((INotifyPropertyChanged)Items).PropertyChanged += new PropertyChangedEventHandler(HandlePropertyChanged);
    } 

//    #endregion Constructors

//    #region Interfaces 

    //------------------------------------------------------ 
    // 
    //  Interfaces
    // 
    //-----------------------------------------------------

//    #region INotifyCollectionChanged

    /// <summary>
    /// CollectionChanged event (per <see cref="INotifyCollectionChanged" />). 
    /// </summary> 
    event NotifyCollectionChangedEventHandler INotifyCollectionChanged.CollectionChanged
    { 
        add     { CollectionChanged += value; }
        remove  { CollectionChanged -= value; }
    }

    /// <summary>
    /// Occurs when the collection changes, either by adding or removing an item. 
    /// </summary> 
    /// <remarks>
    /// see <seealso cref="INotifyCollectionChanged"/> 
    /// </remarks>
//    [field:NonSerializedAttribute()]
    protected /*virtual*/ event NotifyCollectionChangedEventHandler CollectionChanged;

    /// <summary>
    /// raise CollectionChanged event to any listeners 
    /// </summary> 
    protected /*virtual*/ void OnCollectionChanged(NotifyCollectionChangedEventArgs args)
    { 
        if (CollectionChanged != null)
        {
            CollectionChanged(this, args);
        } 
    }

//    #endregion INotifyCollectionChanged 

//    #region INotifyPropertyChanged 

    /// <summary>
    /// PropertyChanged event (per <see cref="INotifyPropertyChanged" />).
    /// </summary> 
    event PropertyChangedEventHandler INotifyPropertyChanged.PropertyChanged
    { 
        add     { PropertyChanged += value; } 
        remove  { PropertyChanged -= value; }
    } 

    /// <summary>
    /// Occurs when a property changes.
    /// </summary> 
    /// <remarks>
    /// see <seealso cref="INotifyPropertyChanged"/> 
    /// </remarks> 
//    [field:NonSerializedAttribute()]
    protected /*virtual*/ event PropertyChangedEventHandler PropertyChanged; 

    /// <summary>
    /// raise PropertyChanged event to any listeners
    /// </summary> 
    protected /*virtual*/ void OnPropertyChanged(PropertyChangedEventArgs args)
    { 
        if (PropertyChanged != null) 
        {
            PropertyChanged(this, args); 
        }
    }

//    #endregion INotifyPropertyChanged 

//    #endregion Interfaces 

//    #region Private Methods

    //------------------------------------------------------
    //
    //  Private Methods
    // 
    //------------------------------------------------------

    // forward CollectionChanged events from the base list to our listeners 
    void HandleCollectionChanged(Object sender, NotifyCollectionChangedEventArgs e)
    { 
        OnCollectionChanged(e);
    }

    // forward PropertyChanged events from the base list to our listeners 
    void HandlePropertyChanged(Object sender, PropertyChangedEventArgs e)
    { 
        OnPropertyChanged(e); 
    }

//    #endregion Private Methods

//    #region Private Fields

    //-----------------------------------------------------
    // 
    //  Private Fields 
    //
    //------------------------------------------------------ 

//    #endregion Private Fields
}

File: summer/WebContent/js/system/collections/objectmodel/ReadOnlyObservableCollection.js
/**
 * ReadOnlyObservableCollection
 */

define(["dojo/_base/declare", "system/Type", "objectmodel/ReadOnlyCollection", "specialized/INotifyCollectionChanged", 
        "specialized/NotifyCollectionChangedEventHandler", "componentmodel/PropertyChangedEventHandler", 
        "componentmodel/INotifyPropertyChanged"], 
		function(declare, Type, ReadOnlyCollection, INotifyCollectionChanged,
				NotifyCollectionChangedEventHandler, PropertyChangedEventHandler, 
				INotifyPropertyChanged){
	var ReadOnlyObservableCollection = declare("ReadOnlyObservableCollection", 
			[ReadOnlyCollection, INotifyCollectionChanged, INotifyPropertyChanged],{
		constructor:function(/*ObservableCollection<T>*/ list) 
	    { 
//			ReadOnlyCollection.prototype.constructor.call(this, list);
	        this.Items.CollectionChanged.Combine(
	        		new NotifyCollectionChangedEventHandler(this, this.HandleCollectionChanged)); 
	        this.Items.PropertyChanged.Combine(new PropertyChangedEventHandler(this, this.HandlePropertyChanged));
	    },
	    
	    /// <summary>
        /// raise CollectionChanged event to any listeners 
        /// </summary> 
//        protected virtual void 
	    OnCollectionChanged:function(/*NotifyCollectionChangedEventArgs*/ args)
        { 
            if (this.CollectionChanged != null)
            {
            	this.CollectionChanged.Invoke(this, args);
            } 
        },
        
        /// <summary>
        /// raise PropertyChanged event to any listeners
        /// </summary> 
//        protected virtual void 
        OnPropertyChanged:function(/*PropertyChangedEventArgs*/ args)
        { 
            if (this.PropertyChanged != null) 
            {
            	this.PropertyChanged.Invoke(this, args); 
            }
        },
 
        // forward CollectionChanged events from the base list to our listeners 
//        void 
        HandleCollectionChanged:function(/*object*/ sender, /*NotifyCollectionChangedEventArgs*/ e)
        { 
            this.OnCollectionChanged(e);
        },

        // forward PropertyChanged events from the base list to our listeners 
//        void 
        HandlePropertyChanged:function(/*object*/ sender, /*PropertyChangedEventArgs*/ e)
        { 
            this.OnPropertyChanged(e); 
        },
	
	});
	
	Object.defineProperties(ReadOnlyObservableCollection.prototype,{
        /// <summary>
        /// Occurs when the collection changes, either by adding or removing an item. 
        /// </summary> 
        /// <remarks>
        /// see <seealso cref="INotifyCollectionChanged"/> 
        /// </remarks>
//        protected virtual event NotifyCollectionChangedEventHandler 
		CollectionChanged:{
			get:function(){
				if(this._CollectionChanged === undefined){
					this._CollectionChanged = new NotifyCollectionChangedEventHandler();
				}
				
				return this._CollectionChanged;
			}

		}, 
        
        /// <summary>
        /// Occurs when a property changes.
        /// </summary> 
        /// <remarks>
        /// see <seealso cref="INotifyPropertyChanged"/> 
        /// </remarks> 
//        protected virtual event PropertyChangedEventHandler 
		PropertyChanged:{
			get:function(){
				if(this._PropertyChanged === undefined){
					this._PropertyChanged = new PropertyChangedEventHandler();
				}
				
				return this._PropertyChanged;
			}
		} 
	});
	
	Object.defineProperties(ReadOnlyObservableCollection,{
		  
	});
	
	ReadOnlyObservableCollection.Type = new Type("ReadOnlyObservableCollection", ReadOnlyObservableCollection,
			[ReadOnlyCollection.Type, INotifyCollectionChanged.Type, INotifyPropertyChanged.Type]);
	return ReadOnlyObservableCollection;
});





 
